Submitty/Submitty
# [Bugfix:TAGrading] PDF Repair

### What is the current behavior?
Fixes #6699 
Fixes #6700 
Currently, various actions on the PDF annotation page can cause PDFs to get into a broken state where either the PDF won't load, or weird behavior occurs such as annotations disappearing.
Also, there is a bug where Submitty is giving graders the student view of the pdf, causing their grading to not save.

### What is the new behavior?
Bugfixes:
* Submitty gives the correct version of the pdf view, so graders' annotations are properly saved.
* If Submitty detects any faulty annotations from the current user upon loading, it will attempt give a prompt to the user asking if they would like to try and repair them.
  * If they accept, Submitty will try to reset properties with null values to a reasonable default value.
  * If no default value is found, the annotation will be deleted.
  * Submitty will save the repaired annotations and refresh the page.
 
![image](https://user-images.githubusercontent.com/28243927/126214428-60a0a278-5de1-46a3-b5ba-112a6f6dcab8.png)
* If the user declines to do a repair, a button will be added to allow them to do it again at any time, and a red danger message will show above the pdf

![image](https://user-images.githubusercontent.com/28243927/126214581-8a0290b9-2fe7-4835-945e-c2d8d59e0eb4.png)

 * If Submitty detects any faulty annotations from NOT the current user, it will alert the current user and temporarily fix them so that the user can make their own annotations.
   * In order to actually repair these annotations, the user who created them must visit the page and agree to repair them. (see above)
 
![image](https://user-images.githubusercontent.com/28243927/126214992-676ee031-aa05-4357-a89e-27a13203676b.png)

Other Improvements: 
* Pen color/size selector menus will close when a click occurs outside of them.
* More checking has been done to help prevent faulty annotations from being created
* Improved speed of pdf toolbar actions
* Allow custom setting of zoom level
  * Zoom level is remembered across different students
* Added rotate left button
* Added Clear button 
  * Only clears your annotations (saves and refreshes page)
* Added button to hide/show other users' annotations
* Allows graders to specify a custom color from a color picker

### Testing
The easiest way to test is to modify the annotations in localStorage.
1. Log in (I will use 'instructor')
2. Go to pdf annotation page.
3. Make a text annotation
4. Open console
5. Enter `let annotations = JSON.parse(localStorage.getItem('<PDF NAME>.pdf/<CURRENT USER>/annotations'))`
6. Make a modification to the annotations array to set a property on an annotation to null. For example if there is only a single text annotation, `a[0].size = null`, or `a[0].content = null`
7. Save the annotations to storage by doing 
`localStorage.setItem('<PDF NAME>.pdf/<CURRENT USER>/annotations', JSON.stringify(annotations))`
8. Press the save button to save the faulty annotations to the server.
9. Refresh the page to see repair popup
